src/falcon/falcon_c/falcon_rpo.c

```diff
@@ -0,0 +1,425 @@
+ * NONCELEN bytes. sigbuf[] receives the signature value (without nonce
+ * or header byte), with *sigbuflen providing the maximum value length and
+ * receiving the actual value length.
+ *
+ * If a signature could be computed but not encoded because it would
+ * exceed the output buffer size, then a new signature is computed. If
+ * the provided buffer size is too low, this could loop indefinitely, so
+ * the caller must provide a size that can accommodate signatures with a
+ * large enough probability.
+ *
+ * Return value: 0 on success, -1 on error.
+ */
+static int
+do_sign(uint8_t *nonce, uint8_t *sigbuf, size_t *sigbuflen,
+        const uint8_t *m, size_t mlen, const uint8_t *sk) {
+    union {
+        uint8_t b[72 * 512];
+        uint64_t dummy_u64;
+        fpr dummy_fpr;
+    } tmp;
+    int8_t f[512], g[512], F[512], G[512];
+    struct {
+        int16_t sig[512];
+        uint16_t hm[512];
+    } r;
+    unsigned char seed[48];
+    inner_shake256_context sc;
+    inner_rpo128_context rc;
+    size_t u, v;
+
+    /*
+     * Decode the private key.
+     */
+    if (sk[0] != 0x50 + 9) {
+        return -1;
+    }
+    u = 1;
+    v = PQCLEAN_FALCON512_CLEAN_trim_i8_decode(
+            f, 9, PQCLEAN_FALCON512_CLEAN_max_fg_bits[9],
+            sk + u, PQCLEAN_FALCON512_CLEAN_CRYPTO_SECRETKEYBYTES - u);
+    if (v == 0) {
+        return -1;
+    }
+    u += v;
+    v = PQCLEAN_FALCON512_CLEAN_trim_i8_decode(
+            g, 9, PQCLEAN_FALCON512_CLEAN_max_fg_bits[9],
+            sk + u, PQCLEAN_FALCON512_CLEAN_CRYPTO_SECRETKEYBYTES - u);
+    if (v == 0) {
+        return -1;
+    }
+    u += v;
+    v = PQCLEAN_FALCON512_CLEAN_trim_i8_decode(
+            F, 9, PQCLEAN_FALCON512_CLEAN_max_FG_bits[9],
+            sk + u, PQCLEAN_FALCON512_CLEAN_CRYPTO_SECRETKEYBYTES - u);
+    if (v == 0) {
+        return -1;
+    }
+    u += v;
+    if (u != PQCLEAN_FALCON512_CLEAN_CRYPTO_SECRETKEYBYTES) {
+        return -1;
+    }
+    if (!PQCLEAN_FALCON512_CLEAN_complete_private(G, f, g, F, 9, tmp.b)) {
+        return -1;
+    }
+
+    /*
+     * Create a random nonce (40 bytes).
+     */
+    randombytes(nonce, NONCELEN);
+
+    // Transform the nonce into 8 chunks each of size 8 bytes. We do this in order to be sure that
+    // the conversion to field elements succeeds
+    uint8_t buffer[64];
+    memset(buffer, 0, 64);
+    for (size_t i = 0; i < 8; i++)
+    {
+        buffer[8*i] = nonce[5*i];
+        buffer[8*i + 1] = nonce[5*i + 1];
+        buffer[8*i + 2] = nonce[5*i + 2];
+        buffer[8*i + 3] = nonce[5*i + 3];
+        buffer[8*i + 4] = nonce[5*i + 4];
+    }
+    
+
+    /*
+     * Hash message nonce + message into a vector.
+     */
+    inner_rpo128_init(&rc);
+    inner_rpo128_inject(&rc, buffer, NONCELEN + 24);
+    inner_rpo128_inject(&rc, m, mlen);
+    inner_rpo128_flip(&rc);
+    PQCLEAN_FALCON512_CLEAN_hash_to_point_rpo(&rc, r.hm, 9);
+    // TODO: should we inner_rpo128_ctx_release(&rc) ?
```

Comment:
I believe this is the only section which has been modified as compared to the original function - so, I'd probably put very clear comments around it to convey this.

Reply:
Done